src/engraving/dom/engravingitem.cpp

```diff
@@ -1638,6 +1638,85 @@ RectF EngravingItem::symBbox(const SymIdList& symbols) const
     return score()->engravingFont()->bbox(symbols, magS());
 }
 
+Shape EngravingItem::symShape(SymId id) const
+{
+    Shape shape;
+
+    RectF boundingBox = symBbox(id);
+    double bottom = boundingBox.bottom();
+    double top = boundingBox.top();
+    double left = boundingBox.left();
+    double right = boundingBox.right();
+
+    PointF cutOutNW = symSmuflAnchor(id, SmuflAnchorId::cutOutNW);
+    PointF cutOutNE = symSmuflAnchor(id, SmuflAnchorId::cutOutNE);
+    PointF cutOutSW = symSmuflAnchor(id, SmuflAnchorId::cutOutSW);
+    PointF cutOutSE = symSmuflAnchor(id, SmuflAnchorId::cutOutSE);
+
+    bool nwNull = cutOutNW.isNull();
+    bool neNull = cutOutNE.isNull();
+    bool swNull = cutOutSW.isNull();
+    bool seNull = cutOutSE.isNull();
+
+    if (nwNull && neNull && swNull && seNull) {
+        return Shape(symBbox(id), this);
+    }
+
+    if (nwNull) {
+        cutOutNW = PointF(left, top);
+    }
+    if (neNull) {
+        cutOutNE = PointF(right, top);
+    }
+    if (swNull) {
+        cutOutSW = PointF(left, bottom);
+    }
+    if (seNull) {
+        cutOutSE = PointF(right, bottom);
+    }
+
+    double leftInset = std::max(cutOutNW.x(), cutOutSW.x());
+    double rightInset = std::min(cutOutNE.x(), cutOutSE.x());
+    double topInset = std::max(cutOutNW.y(), cutOutNE.y());
+    double bottomInset = std::min(cutOutSW.y(), cutOutSE.y());
+
+    std::vector<RectF> rects;
+    rects.reserve(6); //at most
+
+    // bottom rect
+    rects.emplace_back(RectF(PointF(cutOutSW.x(), bottom), PointF(cutOutSE.x(), topInset)).normalized());
+    // right rect
+    bool rightRectPlaced = false;
+    if (!seNull) {
+        rects.emplace_back(RectF(PointF(right, cutOutSE.y()), PointF(leftInset, cutOutNE.y())).normalized());
+        rightRectPlaced = true;
+    }
+    // top rect
+    bool topRectPlaced = false;
+    if (!rightRectPlaced || !neNull) {
+        rects.emplace_back(RectF(PointF(cutOutNW.x(), top), PointF(cutOutNE.x(), bottomInset)).normalized());
+        topRectPlaced = true;
+    }
+    // left rect
+    if (!topRectPlaced || !nwNull) {
+        rects.emplace_back(RectF(PointF(left, cutOutSW.y()), PointF(rightInset, cutOutNW.y())).normalized());
+    }
+    // center horizontal rect if needed
+    if (leftInset > rightInset && topInset < bottomInset) {
+        rects.emplace_back(RectF(PointF(left, bottomInset), PointF(right, topInset)).normalized());
+    }
+    // center vertical rect if needed
+    if (leftInset < rightInset && topInset > bottomInset) {
+        rects.emplace_back(RectF(PointF(leftInset, bottom), PointF(rightInset, top)).normalized());
+    }
+
+    for (RectF& rect : rects) {
+        shape.add(rect, this);
```

Comment:
Maybe we should add method like `addAll(const std::vector<RectF>& rects, EngravingItem* item)`
then we can do m_elements.reserve to avoid relocated several times

Reply:
or we can add method reserve to Shape class :)

Reply:
I made a new Shape constructor for this case :)